File: app/src/main/java/io/github/mattsays/rommnative/model/OfflineModels.kt
package io.github.mattsays.rommnative.model

enum class ConnectivityState {
    ONLINE,
    OFFLINE,
}

enum class OfflineHomeFeedType {
    CONTINUE_PLAYING,
    RECENTLY_ADDED,
    FEATURED_COLLECTION,
}

enum class PendingRemoteActionType {
    GAME_SYNC,
    CORE_DOWNLOAD,
}

enum class PendingRemoteActionStatus {
    PENDING,
    RUNNING,
    FAILED,
}

enum class MediaCacheCategory {
    PLATFORM_LOGO,
    ROM_COVER,
    COLLECTION_COVER,
}

data class OfflineState(
    val connectivity: ConnectivityState = ConnectivityState.ONLINE,
    val activeProfileId: String? = null,
    val catalogReady: Boolean = false,
    val mediaReady: Boolean = false,
    val isRefreshing: Boolean = false,
    val lastFullSyncAtEpochMs: Long? = null,
    val lastMediaSyncAtEpochMs: Long? = null,
    val lastError: String? = null,
    val cacheBytes: Long = 0L,
) {
    val isOffline: Boolean
        get() = connectivity == ConnectivityState.OFFLINE

    val isOfflineReady: Boolean
        get() = catalogReady && mediaReady
}

data class PendingRemoteAction(
    val profileId: String,
    val actionType: PendingRemoteActionType,
    val dedupeKey: String,
    val payloadJson: String,
    val status: PendingRemoteActionStatus,
    val createdAtEpochMs: Long,
    val updatedAtEpochMs: Long,
    val lastError: String? = null,
)

data class PendingGameSyncPayload(
    val romId: Int,
    val fileId: Int,
)

data class PendingCoreDownloadPayload(
    val runtimeId: String,
)

File: ios/Sources/RommioContract/OfflineModels.swift
import Foundation

public enum ConnectivityState: String, Codable, Sendable, CaseIterable {
    case online = "ONLINE"
    case offline = "OFFLINE"
}

public struct OfflineState: Codable, Hashable, Sendable {
    public var connectivity: ConnectivityState
    public var activeProfileID: String?
    public var catalogReady: Bool
    public var mediaReady: Bool
    public var isRefreshing: Bool
    public var lastFullSyncAtEpochMS: Int64?
    public var lastMediaSyncAtEpochMS: Int64?
    public var lastError: String?
    public var cacheBytes: Int64

    public init(
        connectivity: ConnectivityState,
        activeProfileID: String? = nil,
        catalogReady: Bool = false,
        mediaReady: Bool = false,
        isRefreshing: Bool = false,
        lastFullSyncAtEpochMS: Int64? = nil,
        lastMediaSyncAtEpochMS: Int64? = nil,
        lastError: String? = nil,
        cacheBytes: Int64 = 0
    ) {
        self.connectivity = connectivity
        self.activeProfileID = activeProfileID
        self.catalogReady = catalogReady
        self.mediaReady = mediaReady
        self.isRefreshing = isRefreshing
        self.lastFullSyncAtEpochMS = lastFullSyncAtEpochMS
        self.lastMediaSyncAtEpochMS = lastMediaSyncAtEpochMS
        self.lastError = lastError
        self.cacheBytes = cacheBytes
    }

    enum CodingKeys: String, CodingKey {
        case connectivity
        case activeProfileID = "active_profile_id"
        case catalogReady = "catalog_ready"
        case mediaReady = "media_ready"
        case isRefreshing = "is_refreshing"
        case lastFullSyncAtEpochMS = "last_full_sync_at_epoch_ms"
        case lastMediaSyncAtEpochMS = "last_media_sync_at_epoch_ms"
        case lastError = "last_error"
        case cacheBytes = "cache_bytes"
    }
}
